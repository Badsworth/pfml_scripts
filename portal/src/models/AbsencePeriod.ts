import { LeaveReasonType } from "./LeaveReason";
import { groupBy } from "lodash";

export type AbsencePeriodRequestDecision =
  | "Approved"
  | "Cancelled"
  | "Denied"
  | "In Review"
  | "Pending"
  | "Completed"
  // Projected is (at least) used for Bonding Leave periods created automatically
  // for periods with Birth Disability as the reason qualifier
  | "Projected"
  | "Voided"
  | "Withdrawn";

export type AbsencePeriodTypes =
  | "Continuous"
  | "Intermittent"
  | "Reduced Schedule";

export class AbsencePeriod {
  absence_period_end_date: string;
  absence_period_start_date: string;
  fineos_leave_request_id: string | null = null;
  period_type: AbsencePeriodTypes;
  reason: LeaveReasonType;
  reason_qualifier_one = "";
  reason_qualifier_two = "";
  request_decision: AbsencePeriodRequestDecision;

  constructor(attrs: Partial<AbsencePeriod> = {}) {
    Object.assign(this, attrs);
  }

  static groupByReason(absence_periods: AbsencePeriod[]): {
    [reason: string]: AbsencePeriod[];
  } {
    return groupBy(absence_periods, "reason");
  }

  /**
   * @returns periods sorted newest to oldest (by start date)
   */
  static sortNewToOld(absence_periods: AbsencePeriod[]): AbsencePeriod[] {
    const periods = absence_periods.slice(); // avoids mutating the original array
    return periods.sort((a, b) => {
      return a.absence_period_start_date > b.absence_period_start_date ? -1 : 1;
    });
  }
}
