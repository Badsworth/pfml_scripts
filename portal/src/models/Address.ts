/* eslint sort-keys: ["error", "asc"] */
import BaseModel from "./BaseModel";

class Address extends BaseModel {
  // @ts-expect-error ts-migrate(2416) FIXME: Property 'defaults' in type 'Address' is not assig... Remove this comment to see the full error message
  get defaults() {
    return {
      city: null,
      line_1: null,
      line_2: null,
      state: null,
      zip: null,
    };
  }
}

export default Address;
