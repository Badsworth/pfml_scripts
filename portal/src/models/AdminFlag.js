import BaseModel from "./BaseModel";

class FlagModel extends BaseModel {
  get defaults() {
    return {
      flag_id: null,
      name: null,
      enabled: false,
      options: {},
      start: null,
      end: null,
    };
  }
}

export default FlagModel;
