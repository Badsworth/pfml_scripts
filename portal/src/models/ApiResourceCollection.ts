export interface ApiResource {
  // API resources are expected to be objects. We expect at least an ID property.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any;
}

/**
 * Readonly generic class representing a collection of a specific data model. Methods
 * return new instances of the collection rather than modifying the original collection.
 * @example const c = new ApiResourceCollection<Document>("document_id", [{ document_id: "123", name: "example" }]);
 *          c.setItem({ document_id: "456", name: "example2" });
 */
class ApiResourceCollection<TApiResource extends ApiResource> {
  idKey: keyof TApiResource;
  private map: Map<string, TApiResource>;

  constructor(idKey: keyof TApiResource, items: TApiResource[] = []) {
    this.idKey = idKey;
    this.map = new Map(
      items.map((item) => {
        return [item[idKey], item];
      })
    );
  }

  private get _mutableMap() {
    return new Map(this.map);
  }

  private _arrayFromMap(map: Map<string, TApiResource>): TApiResource[] {
    return Array.from(map.values());
  }

  get items(): TApiResource[] {
    return this._arrayFromMap(this.map);
  }

  get isEmpty(): boolean {
    return this.map.size === 0;
  }

  getItem(itemId: string): TApiResource | undefined {
    return this.map.get(itemId);
  }

  /**
   * Add or update an item in the collection. Returns a new collection with the
   * item included. Does not modify the original collection.
   */
  setItem(item: TApiResource): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap.set(item[this.idKey], item);
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Add or update items in the collection. Returns a new collection with the
   * items included. Does not modify the original collection.
   */
  setItems(items: TApiResource[] = []): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    items.forEach((item) => {
      updatedMap.set(item[this.idKey], item);
    });
    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }

  /**
   * Removes an item from the collection. Returns a new collection with the item removed.
   * Does not modify the original collection.
   */
  removeItem(itemId: string): ApiResourceCollection<TApiResource> {
    const updatedMap = this._mutableMap;
    updatedMap.delete(itemId);

    return new ApiResourceCollection<TApiResource>(
      this.idKey,
      this._arrayFromMap(updatedMap)
    );
  }
}

export default ApiResourceCollection;
