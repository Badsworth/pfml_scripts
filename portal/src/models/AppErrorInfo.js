import BaseModel from "./BaseModel";
import uniqueId from "lodash/uniqueId";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property {string} [field] - Path of the field the error is associated with
 * @property {string} key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property {string} message - Internationalized message displayed to the user (like `Error#message`)
 * @property {string} [name] - Name of the error (like `Error#name`)
 */
class AppErrorInfo extends BaseModel {
  get defaults() {
    return {
      key: uniqueId("AppErrorInfo"),
      message: null,
      name: null,
      field: null,
    };
  }
}

export default AppErrorInfo;
