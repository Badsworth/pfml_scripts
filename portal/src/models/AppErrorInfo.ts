import { uniqueId } from "lodash";

/**
 * Provides a consistent interface for creating error messages displayed to the user.
 * @property [field] - Path of the field the error is associated with
 * @property key - Unique key used for React `key` prop (https://reactjs.org/docs/lists-and-keys.html)
 * @property message - Internationalized message displayed to the user (like `Error#message`)
 * @property [name] - Name of the error (like `Error#name`)
 * @property [meta] - Additional error data, like the application_id associated with it
 * @property [rule] - Name of validation issue rule (e.g "min_leave_periods", "conditional", etc)
 * @property [type] - Name of validation issue type (e.g "required", "pattern", "date", etc)
 */
class AppErrorInfo {
  key: string = uniqueId("AppErrorInfo");
  message?: string | JSX.Element;
  name?: string;
  field?: string;
  meta?: Record<string, unknown>;
  rule?: string;
  type?: string;

  constructor(attrs: Partial<AppErrorInfo>) {
    Object.assign(this, attrs);
  }
}

export default AppErrorInfo;
