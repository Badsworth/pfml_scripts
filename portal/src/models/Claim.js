/* eslint sort-keys: ["error", "asc"] */
/**
 * @file Benefits application model and enum values
 */
import BaseModel from "./BaseModel";

class Claim extends BaseModel {
  get defaults() {
    return {
      application_id: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      avg_weekly_hours_worked: null,
      created_at: null,
      duration_type: null,
      employee_ssn: null,
      employer_benefits: [],
      employer_fein: null,
      // TODO: This field does not yet exist in the API: https://lwd.atlassian.net/browse/CP-574
      employment_status: null,
      first_name: null,
      // TODO: We'll map this to the correct API field once we get into Intermittent Leave work
      hours_off_needed: null,
      last_name: null,
      leave_details: {
        continuous_leave_periods: null,
        employer_notification_date: null,
        employer_notified: null,
        reason: null,
      },
      middle_name: null,
      other_incomes: [],
      pregnant_or_recent_birth: null,
      previous_leaves: [],
      status: null,
    };
  }
}

/**
 * Enums for the Application's `status` field
 * @enum {string}
 */
export const ClaimStatus = {
  // Submitted by the user (Step 2)
  completed: "Completed",
  // AKA: In Progress (Step 1)
  started: "Started",
  // Stored in the claims processing system (Step 3)
  submitted: "Submitted",
};

/**
 * Enums for the Application's `leave_details.employment_status` field
 * @enum {string}
 */
export const EmploymentStatus = {
  employed: "employed",
  selfEmployed: "self-employed",
  unemployed: "unemployed",
};

/**
 * Enums for the Application's `leave_details.reason` field
 * @enum {string}
 */
export const LeaveReason = {
  // TODO: We need to map some of these to the correct API fields,
  // once those enum values exist within the API
  // https://lwd.atlassian.net/browse/CP-515
  activeDutyFamily: "Care For A Family Member",
  bonding: "Child Bonding",
  medical: "Serious Health Condition - Employee",
  serviceMemberFamily: "Pregnancy/Maternity",
};

export default Claim;
