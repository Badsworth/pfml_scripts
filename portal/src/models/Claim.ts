/* eslint sort-keys: ["error", "asc"] */
import { compact, get } from "lodash";

/**
 * A record from the API's Claims table. Could be utilized by Leave Admin and Claimants.
 */
class Claim {
  absence_period_end_date: string | null = null;
  absence_period_start_date: string | null = null;
  claim_status: AbsenceCaseStatusType | null;
  claim_type_description: string | null = null;
  created_at: string;
  employee: ClaimEmployee | null;
  employer: ClaimEmployer;
  fineos_absence_id: string;
  fineos_notification_id: string;
  managed_requirements: ManagedRequirement[];

  constructor(attrs: Claim) {
    Object.assign(this, attrs);
    if (attrs.employee) {
      this.employee = new ClaimEmployee(attrs.employee);
    }
  }
}

/**
 * Employee (claimant) record associated to the Claim
 */
export class ClaimEmployee {
  first_name: string | null = null;
  last_name: string | null = null;
  middle_name: string | null = null;
  other_name: string | null = null;

  constructor(attrs: Partial<ClaimEmployee>) {
    Object.assign(this, attrs);
  }

  get fullName() {
    return compact([
      get(this, "first_name"),
      get(this, "middle_name"),
      get(this, "last_name"),
    ]).join(" ");
  }
}

/**
 * Employer record associated to the Claim
 */
export interface ClaimEmployer {
  employer_dba: string | null;
  employer_fein: string;
  employer_id: string;
}

/**
 * Managed requirements associated to the Claim
 */
export interface ManagedRequirement {
  category: string;
  created_at: string;
  follow_up_date: string | null;
  responded_at: string | null;
  status: "Open" | "Complete" | "Suppressed";
  type: string;
}

/**
 * Enums for `claim.claim_status` field
 * This is the subset of potential values returned by the API that we show to the user.
 * We currently ignore other potential values like "Adjudication", "Intake In Progress", "Unknown"
 * @enum {string}
 */
export const AbsenceCaseStatus = {
  approved: "Approved",
  closed: "Closed",
  completed: "Completed",
  declined: "Declined",
} as const;

export type AbsenceCaseStatusType =
  typeof AbsenceCaseStatus[keyof typeof AbsenceCaseStatus];

export default Claim;
