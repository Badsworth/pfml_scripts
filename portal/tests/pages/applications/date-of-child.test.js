import {
  MockClaimBuilder,
  renderWithAppLogic,
  simulateEvents,
} from "../../test-utils";
import DateOfChild from "../../../src/pages/applications/date-of-child";
import { DateTime } from "luxon";
import { ReasonQualifier } from "../../../src/models/Claim";

jest.mock("../../../src/hooks/useAppLogic");

const past = DateTime.local().minus({ months: 1 }).toISODate();
const now = DateTime.local().toISODate();
const future = DateTime.local().plus({ months: 1 }).toISODate();

const child_birth_date = "leave_details.child_birth_date";
const child_placement_date = "leave_details.child_placement_date";

const setup = (claimAttrs) => {
  const { appLogic, claim, wrapper } = renderWithAppLogic(DateOfChild, {
    claimAttrs,
  });

  const { changeField, submitForm } = simulateEvents(wrapper);

  return {
    appLogic,
    changeField,
    claim,
    submitForm,
    wrapper,
  };
};

describe("DateOfChild", () => {
  it("renders the page", () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it("shows the birth date input with hint text when the claim is for a newborn", () => {
    const claim = new MockClaimBuilder().bondingBirthLeaveReason().create();
    const { wrapper } = setup(claim);

    expect(wrapper.find({ name: child_birth_date }).exists()).toBeTruthy();
    expect(wrapper.find({ name: child_placement_date }).exists()).toBeFalsy();
    expect(wrapper.find("InputDate").prop("hint")).toMatchInlineSnapshot(
      `"If your child has not been born yet, enter the expected due date."`
    );
  });

  it("shows the correct input question when the claim is for an adoption", () => {
    const claim = new MockClaimBuilder().bondingAdoptionLeaveReason().create();
    const { wrapper } = setup(claim);

    expect(wrapper.find({ name: child_placement_date }).exists()).toBeTruthy();
    expect(wrapper.find({ name: child_birth_date }).exists()).toBeFalsy();
  });

  describe("when child birth or placement date is in the future", () => {
    let spy;

    beforeAll(() => {
      spy = jest.spyOn(DateTime, "local").mockImplementation(() => {
        return {
          toISODate: () => now,
        };
      });
    });

    afterAll(() => {
      spy.mockRestore();
    });

    it("sets has_future_child_date as true for future birth bonding leave when claim is already populated", async () => {
      const claim = new MockClaimBuilder()
        .bondingBirthLeaveReason(future)
        .create();
      const { appLogic, submitForm } = setup(claim);

      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: future,
          child_placement_date: null,
          has_future_child_date: true,
        },
      });
    });

    it("sets has_future_child_date as true for future birth bonding leave when data is manually populated", async () => {
      const claim = new MockClaimBuilder().bondingLeaveReason().create();
      claim.leave_details.reason_qualifier = ReasonQualifier.newBorn;

      const { appLogic, changeField, submitForm } = setup(claim);

      changeField("leave_details.child_birth_date", future);

      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: future,
          child_placement_date: null,
          has_future_child_date: true,
        },
      });
    });

    it("sets has_future_child_date as true for future placement bonding leave when claim is already populated", async () => {
      const claim = new MockClaimBuilder()
        .bondingFosterCareLeaveReason(future)
        .create();
      const { appLogic, submitForm } = setup(claim);

      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: null,
          child_placement_date: future,
          has_future_child_date: true,
        },
      });
    });

    it("sets has_future_child_date as true for future placement bonding leave when data is manually populated", async () => {
      const claim = new MockClaimBuilder().bondingLeaveReason().create();
      claim.leave_details.reason_qualifier = ReasonQualifier.fosterCare;

      const { appLogic, changeField, submitForm } = setup(claim);
      changeField("leave_details.child_placement_date", future);
      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: null,
          child_placement_date: future,
          has_future_child_date: true,
        },
      });
    });
  });

  describe("when child birth or placement date is in the past", () => {
    let spy;

    beforeAll(() => {
      spy = jest.spyOn(DateTime, "local").mockImplementation(() => {
        return {
          toISODate: () => now,
        };
      });
    });

    afterAll(() => {
      spy.mockRestore();
    });

    it("sets has_future_child_date as false for past birth bonding leave when claim is pre-populated", async () => {
      const claim = new MockClaimBuilder()
        .bondingBirthLeaveReason(past)
        .create();
      const { appLogic, submitForm } = setup(claim);

      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: past,
          child_placement_date: null,
          has_future_child_date: false,
        },
      });
    });

    it("sets has_future_child_date as false for past birth bonding leave when data is manually populated", async () => {
      const claim = new MockClaimBuilder().bondingLeaveReason().create();
      claim.leave_details.reason_qualifier = ReasonQualifier.newBorn;

      const { appLogic, changeField, submitForm } = setup(claim);
      changeField("leave_details.child_birth_date", past);
      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: past,
          child_placement_date: null,
          has_future_child_date: false,
        },
      });
    });

    it("sets has_future_child_date as false for past placement bonding leave when claim is pre-populated", async () => {
      const claim = new MockClaimBuilder()
        .bondingFosterCareLeaveReason(past)
        .create();
      const { appLogic, submitForm } = setup(claim);
      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: null,
          child_placement_date: past,
          has_future_child_date: false,
        },
      });
    });

    it("sets has_future_child_date as false for past placement bonding leave when data is manually populated", async () => {
      const claim = new MockClaimBuilder().bondingLeaveReason().create();
      claim.leave_details.reason_qualifier = ReasonQualifier.fosterCare;

      const { appLogic, changeField, submitForm } = setup(claim);
      changeField("leave_details.child_placement_date", past);
      await submitForm();

      expect(appLogic.claims.update).toHaveBeenCalledWith(expect.any(String), {
        leave_details: {
          child_birth_date: null,
          child_placement_date: past,
          has_future_child_date: false,
        },
      });
    });
  });
});
