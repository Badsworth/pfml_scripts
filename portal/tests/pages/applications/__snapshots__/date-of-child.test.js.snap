// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateOfChild renders the page 1`] = `
<div>
  <form
    class="usa-form"
    method="post"
  >
    <h1
      class="js-title margin-top-0 margin-bottom-2 font-heading-sm line-height-sans-3"
      tabindex="-1"
    >
      Leave details
    </h1>
    <fieldset
      class="usa-fieldset margin-top-3 usa-form-group"
    >
      <legend
        class="usa-label text-bold usa-legend font-heading-lg line-height-sans-3 margin-bottom-1 maxw-tablet"
      >
        pages.claimsBondingDateOfChild.sectionLabel
      </legend>
      <span
        class="display-block line-height-sans-5 usa-hint text-base-dark measure-5"
      >
        MM / DD / YYYY
      </span>
      <div
        class="usa-memorable-date"
      >
        <div
          class="usa-form-group usa-form-group--month"
        >
          <label
            class="usa-label text-normal margin-top-05 font-heading-xs measure-5"
            for="InputText1"
            id="InputText1_label"
          >
            Month
          </label>
          <input
            aria-labelledby="InputText1_label InputText1_hint InputText1_error"
            class="usa-input usa-input--inline maxw-mobile-lg"
            id="InputText1"
            inputmode="numeric"
            maxlength="2"
            name="leave_details.child_placement_date_month"
            type="text"
            value=""
          />
        </div>
        <div
          class="usa-form-group usa-form-group--day"
        >
          <label
            class="usa-label text-normal margin-top-05 font-heading-xs measure-5"
            for="InputText2"
            id="InputText2_label"
          >
            Day
          </label>
          <input
            aria-labelledby="InputText2_label InputText2_hint InputText2_error"
            class="usa-input usa-input--inline maxw-mobile-lg"
            id="InputText2"
            inputmode="numeric"
            maxlength="2"
            name="leave_details.child_placement_date_day"
            type="text"
            value=""
          />
        </div>
        <div
          class="usa-form-group usa-form-group--year"
        >
          <label
            class="usa-label text-normal margin-top-05 font-heading-xs measure-5"
            for="InputText3"
            id="InputText3_label"
          >
            Year
          </label>
          <input
            aria-labelledby="InputText3_label InputText3_hint InputText3_error"
            class="usa-input usa-input--inline maxw-mobile-lg"
            id="InputText3"
            inputmode="numeric"
            maxlength="4"
            name="leave_details.child_placement_date_year"
            type="text"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <button
      class="usa-button position-relative margin-top-4"
      type="submit"
    >
      Save and continue
    </button>
  </form>
</div>
`;
